File: pycode/QuatMath/QuatNormal.py
"""
QuatNormal.py

Normalize a quaternion
Original MATLAB sensor fusion algorithms converted for Python/NumPy
"""

import numpy as np


def quat_normalize(q):
    """
    Normalize a quaternion to unit length
    Enforces q[0] >= 0 for consistent sign convention (matches C implementation)

    Args:
        q (array): Quaternion [w, x, y, z]

    Returns:
        array: Normalized quaternion with q[0] >= 0
    """
    q = np.array(q)
    norm = np.linalg.norm(q)
    if norm == 0:
        return np.array([1, 0, 0, 0])  # Return identity quaternion

    q_norm = q / norm

    # Enforce q[0] >= 0 (matches C QuatNormal function lines 152-156)
    # If q0 is negative, flip all components (q and -q represent same rotation)
    if q_norm[0] < 0.0:
        q_norm = -q_norm

    return q_norm


# Alias for MATLAB compatibility
QuatNormal = quat_normalize

File: matlab/QuatMath/QuatNormal.m
function qnorm = QuatNormal(q)
%QuatNormalize Normalize a quaternion so that its mag = 1
%
%   qnorm = QuatNormal(q)
%   qnorm = q/qmag
%
%   If q(0) is negative, it makes it positive without changing the
%   functionality of quaternion

    qmag = norm(q);
    qnorm = [1; 0; 0; 0];
    if( qmag>0 )
        qnorm = q./qmag;
    end
    if( qnorm(1) < 0 )
        qnorm = -qnorm;
    end
    
end
